power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp

```diff
@@ -61,31 +79,22 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         size_t index;
     };
 
-    template <class T, class U> struct component_list_generator_impl;
-    template <class... C, size_t... index>
-    struct component_list_generator_impl<ComponentList<C...>, std::index_sequence<index...>> {
-        using AllTypes = ComponentList<C...>;
-        static constexpr std::array component_index_map{ComponentEntry{C::name, index}...};
-        static constexpr size_t n_types = sizeof...(C);
-
-        static size_t find_index(std::string const& name) {
-            auto const found = std::ranges::find_if(component_index_map, [&name](auto x) { return x == name; });
-            assert(found != component_index_map.cend());
-            return found->index;
-        }
-
-        template <class Comp> static constexpr size_t index_of() { return container_impl::get_cls_pos_v<Comp, C...>; }
-    };
-    using AllComponents = component_list_generator_impl<ComponentList<ComponentType...>,
-                                                        std::make_index_sequence<sizeof...(ComponentType)>>;
-    static constexpr size_t n_types = AllComponents::n_types;
+    static constexpr size_t n_types = sizeof...(ComponentType);
 
     using SequenceIdx = std::array<std::vector<Idx2D>, n_types>;
 
     using OwnedUpdateDataset = std::tuple<std::vector<typename ComponentType::UpdateType>...>;
 
     static constexpr Idx ignore_output{-1};
 
+    // run functors with all component types
+    template <class Functor> static constexpr void run_functor_with_all_types_return_void(Functor functor) {
+        (functor.template operator()<ComponentType>(), ...);
+    }
+    template <class Functor> static constexpr auto run_functor_with_all_types_return_array(Functor functor) {
+        return std::array{functor.template operator()<ComponentType>()...};
+    }
+
```

Comment:
i feel like this is a utility we may want in `common`

Reply:
You can't. The `ComponentType` is part of the parent variadic argument of `MainModelImpl`. You cannot move it out.